qdrant-landing/content/articles/sayt.md

```diff
@@ -0,0 +1,122 @@
+    let (tokenizer, session, qdrant) = context.get_ref();
+    let (token_ids, attention, type_ids) = tokenize(tokenizer, q);
+    let vector = embed(session, token_ids, attention, type_ids);
+    let filter = section.is_empty().not().then(
+        || Filter::all(Condition::matches("sections", section)));
+    match qdrant
+        .search_points(&SearchPoints {
+            collection_name: COLLECTION_NAME.to_string(),
+            vector,
+            filter,
+            limit: 5,
+            with_payload: Some(true.into()),
+            ..Default::default,
+        }).await
+    {
+        Ok(SearchResponse { result, .. }) => result_json(result),
+        Err(e) => return HttpResponse::InternalServerError().body(e.to_string()),
+    }
+}
+```
+
+Even with avoiding a network roundtrip, the embedding still takes some time. As always in optimization, if you cannot do the work faster, a good solution is to avoid work altogether (please don't tell my employer). This can be done by pre-computing common prefixes and calculating embeddings for them, then storing them in a `prefix_cache` collection. With the [`recommend`](https://docs.rs/qdrant-client/latest/qdrant_client/client/struct.QdrantClient.html#method.recommend) API method, we can now find the best matches without doing any embedding. For now, I use short (up to and including 5 letters) prefixes, but we can also parse the logs to get the most common search terms and add them to the cache, too.
+
+![Qdrant Recommendation](/articles_data/sayt/Qdrant_Recommendation.png)
+
+The code path for this one is simply:
+
+```rust
+match qdrant
+    .recommend(&RecommendPoints {
+        collection_name: COLLECTION_NAME.to_string(),
+        positive: vec![prefix_to_id(q)],
+        limit: 5,
+        with_payload: Some(true.into()),
+        lookup_from: Some(LookupLocation {
+            collection_name: "prefix_cache".to_string(),
+            vector_name: None,
+        }),
+        ..Default::default()
+    })
+    .await
+{
+    Ok(RecommendResponse { result, .. }) => return Ok(result_json(result)),
+    Err(e) => (), // go on to normal semantic search
+}
+```
+
+To make that work, we need to set up the `prefix_cache` collection with points that have the prefix as their `point_id` and the embedding as their `vector`, which lets us do the lookup with no search or index. The `prefix_to_id` function currently uses the `u64` variant of `PointId`, which can hold 8 bytes, enough for this use:
+
+```rust
+fn prefix_to_id(prefix: &str) -> PointId {
+    let len = prefix.len();
+    PointId::from(u64::from_le_bytes(if len < 8 {
+        let mut result = [0_u8; 8];
+        result[..len].copy_from_slice(prefix.as_bytes());
+        result
+    } else {
+        prefix.as_bytes()[..8].try_into().unwrap()
+    }))
+}
+```
+
+If the need arises, one could instead encode the names as UUID, which needs to time-consumingly hash the input, so I decided against this for now.
```

Comment:
> which needs to time-consumingly hash the input

Is it really time-consuming? It's nanoseconds

Reply:
Depends on the hash function. But I agree, even if it's a few 10s or even 100s of nanos, it's unlikely to show up in the profile.